Add width selector for button block

stacimc opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
Issue #24705

Adds a width selector to the sidebar for the Button block, allowing the user to set their button to 25%, 50%, 75%, or 100% of the parent container.

By default, no percentage width is selected and the button width is determined as normal by the size of its content.

## Notes
_This is an alternative approach to #26045, which implements a width selector on the Button block using a new block support.__ This is related to issue #26368.

## How has this been tested?
Manually with the button block, using left, right, and center alignments.

1. Add a `Buttons` block to a post with one or more `Button`s
2. Select a `Button`
3. You should see a Width Settings section under the Inspector Controls, with options for 25, 50, 75, 100%.
4. Toggle a width option and verify that the correct inline style is applied to the block.
5. Save the post and view on the frontend.
6. Verify that the button is resized correctly in the saved post.

## Screenshots <!-- if applicable -->


![button_resize](https://user-images.githubusercontent.com/63313398/97051892-e0dd1500-1534-11eb-9502-afe3093d77e6.gif)


<img width="605" alt="Screen Shot 2020-10-09 at 1 58 03 PM" src="https://user-images.githubusercontent.com/63313398/95630894-82de0700-0a37-11eb-88d0-257aaa80570d.png">


## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

Commit:
Add sidebar width selector for Button block

Initial commit adding a width option for the Button block with
options at 25, 50, 75, 100% widths. By default none are applied
and the button is sized by its content.

Commit:
Persist width changes

Commit:
Apply styling through class name rather than inline

glendaviesnz commented:
This tested well for me. As mentioned in discussions when using right align this can end up slightly odd, but I think this is because of the use of `float:right`. When adding width settings to on a button in another project [we removed the default right float](https://github.com/Automattic/jetpack/pull/17308/files#diff-d98c55c2161bba38b232c18ac2e6ef84R41). @apeatling, I am thinking that it wouldn't be as easy to do that here are existing content may be relying on the existing `float:right` behaviour?

aaronrobertshaw commented:
Code looks good ✨ 

Apart from the align left/right issues, this tests well for me as well.  I do wonder though if this is another opportunity for adding block support for custom widths so other blocks can opt-in and leverage this? This would also allow width options to be configured via themes.

Regarding the alignment issues, I'd vote for them to be addressed in a separate PR. There might be a few things we could do to smooth some of the more common rough edges created by the floated block losing the parent width to calculate the percentage widths from.

Commit:
Clean up CSS

stacimc commented:
> As mentioned in discussions when using right align things can end up slightly odd, but I think this is because of the use of `float:right`.

This is something I had a lot of difficulty with -- I have a version which I can push up for testing where right-aligning the Buttons block causes the individual Buttons to line up flush to the right side of the container when the post is saved:

<img width="319" alt="Screen Shot 2020-10-12 at 10 20 15 AM" src="https://user-images.githubusercontent.com/63313398/95774057-8c07e780-0c74-11eb-9214-d948b8c5908f.png">

This seems better to me because it is more accurate to what you see in the editor and it's what I would assume should happen when you right-align the parent. The CSS for that removes the `float:right` like you said and adds `display: flex` to the Buttons when saved.

> here are existing content may be relying on the existing float:right behaviour

I think that's accurate. The reason I had trouble with this is because I couldn't figure out how to best emulate the right-align behavior in production, which does not necessarily keep the buttons flush to the right side of the container. For example, this happens on production when I right-align a group of buttons within a column (column highlighted in blue, sorry it's hard to see):

<img width="139" alt="Screen Shot 2020-10-12 at 10 16 51 AM" src="https://user-images.githubusercontent.com/63313398/95774357-1e0ff000-0c75-11eb-93eb-579eb606c39c.png">

Commit:
Allow toggling a custom width off

A width selection can now be toggled off by clicking the already
selected option.

apeatling commented:
@mcsf Pinging you to see if this PR looks better since it's adding width specific to the button block. 👍

talldan commented:
There's another PR that changes how alignment works on the buttons block, so it'd probably worth testing against that as well - https://github.com/WordPress/gutenberg/pull/23168

mcsf commented on packages/block-library/src/button/block.json:
```diff
@@ -52,6 +52,9 @@
 			"type": "string"
+		},
+		"width": {
+			"type": "number"
```
Thinking about possible futures for this block, is this the best data representation? How would we express custom width values without introducing compatibility issues with the current schema? That is, this new `width` attribute is not any `number` but actually a percentage represented as `0 <= n <= 100`. Would it be better if it were a real number _x_ `0 <= x <= 1` to distinguish from full pixel values like `width: 200`? Would it be better if termed `widthPercentage`?

All of these are open questions — I don't have a formed opinion on this.

stacimc replied:
Another possibility here would be to add an additional `widthUnit` attribute if we decide to introduce custom width as pixel values. The `search` block takes that approach.

apeatling replied:
Yes this is how the search block handles this, as well as the cover block. I think this approach is future proof based on that.

apeatling replied:
This is the approach the search and cover blocks take, so this should be future proof for adding a unit in the future.

mcsf reviewed (commented):
Looks good overall (I left some questions on the attribute schema to make sure we've given this a bit of thought), but I'd have someone else test and review the style changes.

stacimc committed:
Update InspectorControls to use sentence case

Co-authored-by: Miguel Fonseca <miguelcsf@gmail.com>

Commit:
Remove max width from buttons with custom width set

PR #23168 overhauls alignment controls on the Buttons container and
allows for full width. For a Button inside a very large container,
in order to be set correctly to 100% it must be able to exceed any
configured max-width.

This commit does not remove the max-width for any button that does
NOT have a custom width percentage selected. This is so that
default buttons continue to be sized as they normally would.

stacimc commented:
Thanks for the heads up about PR #23168, @talldan. I pushed up a small change to allow for the larger percentage selections in the new full-width Buttons container.

apeatling reviewed (approved):
This looks good and tests well. I think this is a great start to width support that could be extended in the future to support custom widths and different units.

Commit:
Apply `has-custom-width` class on frontend

Fixes a bug where the inner button was not being set correctly to 100%
width on the frontend, due to the missing classname.`

glendaviesnz reviewed (approved):
This tested well for me.

apeatling commented:
Going to move ahead with this once since it has approval already. If we need to make further changes the we can open follow up PRs.

kebbet commented:
Is there a way to disable this feature?
